src/nfa/builder.cc

```diff
@@ -217,6 +219,48 @@ Nfa builder::create_sigma_star_nfa(mata::Alphabet* alphabet) {
     return nfa;
 }
 
+Nfa builder::create_random_nfa_tabakov_vardi(const size_t num_of_states, const size_t alphabet_size, const float transition_density, const float final_state_density) {
+    if (transition_density < 0 || static_cast<size_t>(transition_density) > num_of_states) {
+        // Maximum of num_of_states^2 unique transitions for one symbol can be created.
+        throw std::runtime_error("Transition density must be in range [0, num_of_states]");
+    }
+    if (final_state_density < 0 || final_state_density > 1) {
+        // Maximum of num_of_states final states can be created.
+        throw std::runtime_error("Final state density must be in range [0, 1]");
+    }
+
+    Nfa nfa{ num_of_states, StateSet{ 0 }, StateSet{ 0 }, new OnTheFlyAlphabet{} };
+
+    // Initialize the random number generator
+    std::random_device rd;  // Seed for the random number engine
+    std::mt19937 gen(rd()); // Mersenne Twister engine
+    std::uniform_int_distribution<State> state_rand_dis(0, num_of_states - 1);
+    std::uniform_int_distribution<Symbol> symbol_rand_dis(0, static_cast<Symbol>(alphabet_size - 1));
+
+    // Create final states
+    const size_t num_of_final_states{ static_cast<size_t>(std::round(static_cast<float>(num_of_states) * final_state_density)) };
+    while (nfa.final.size() < num_of_final_states) {
+        nfa.final.insert(state_rand_dis(gen));
+    }
+
+    // Create transitions
+    const size_t num_of_transitions_per_symbol{ static_cast<size_t>(std::round(static_cast<float>(num_of_states) * transition_density)) };
```

Comment:
According to this it seems that `transition_density` is in [0,1] interval, right?

Reply:
Again, the name `transition_density` I've taken from the literature is not the best.  The constant `num_of_transition_per_symbol` represents a single symbol's total number of transitions within the whole automaton. Therefore, the value can reach `num_of_states^2` because, for each symbol, there can be a transition between each pair of states.

Reply:
Ah ok, I get it now.